gossip/tx_scrambler.go

```diff
@@ -0,0 +1,156 @@
+package gossip
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/bitutil"
+	"slices"
+)
+
+type scramblerEntry struct {
+	hash   common.Hash
+	sender common.Address
+	nonce  uint64
+}
+
+type scramblerSortFunc func(entries []*scramblerEntry, seed []byte) []*scramblerEntry
+
+// getExecutionOrder first removes any entries with duplicate hashes, then sorts the list by XORed hashes.
+// Furthermore, if there are entries with same sender, these entries are sorted by their nonce (lower comes first).
+func getExecutionOrder(entries []*scramblerEntry, sortFunc scramblerSortFunc) []*scramblerEntry {
```

Comment:
This implementation does not follow the specification in the doc. In particular, it is not deterministic. This test fails:
```
func TestGetExecutionOrder_SortIsDeterministic2(t *testing.T) {
	entryA := &scramblerEntry{
		hash: common.Hash{1},
	}
	entryB := &scramblerEntry{
		hash: common.Hash{2},
	}

	res1 := getExecutionOrder([]*scramblerEntry{entryA, entryB}, builtInSort)
	res2 := getExecutionOrder([]*scramblerEntry{entryB, entryA}, builtInSort)

	if !reflect.DeepEqual(res1, res2) {
		t.Fatal("slices have different order - algorithm is not deterministic")
	}
}
```